src/vizier/services/query_broker/script_runner/config_map_source.go

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package scriptrunner
+
+import (
+	"context"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	clientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"px.dev/pixie/src/shared/cvmsgspb"
+	"px.dev/pixie/src/utils"
+)
+
+// ConfigMapSource pulls cron scripts from config maps.
+type ConfigMapSource struct {
+	stop   func()
+	client clientv1.ConfigMapInterface
+}
+
+// NewConfigMapSource constructs a [Source] that extracts cron scripts from config maps with the label "purpose=cron-script".
+// Each config map must contain
+//   - a script.pxl with the pixel script
+//   - a configs.yaml which will be stored in the Configs field of [cvmsgspb.CronScript]
+//   - a cron.yaml that contains a "frequency_s" key
+func NewConfigMapSource(client clientv1.ConfigMapInterface) *ConfigMapSource {
+	return &ConfigMapSource{client: client}
+}
+
+// Start watches for updates to matching configmaps and sends resulting updates on updatesCh.
+func (source *ConfigMapSource) Start(baseCtx context.Context, updatesCh chan<- *cvmsgspb.CronScriptUpdate) (map[string]*cvmsgspb.CronScript, error) {
+	options := metav1.ListOptions{LabelSelector: "purpose=cron-script"}
+	watcher, err := source.client.Watch(baseCtx, options)
```

Comment:
Sorry, just realized I should have called this our earlier. But we switched our k8s API watchers in the other parts of our code to use the informer model, since we've seen that it handles retries/disconnects much better. Here's an example of how we set up our informers: https://github.com/pixie-io/pixie/blob/4ee33417a5a839417f02e0faa8dada2ecee21a1d/src/vizier/services/metadata/controllers/k8smeta/k8s_metadata_utils.go#L77

Reply:
We learned that adding a `ResourceEventHandler` to the informer produces add events for every initial item in the list (regardless of when you add the event handler). In an effort to satisfy the API contract of `Source`, we needed to discard the add events for the initial items in the list so that we can return the initial scripts and only send changes to that initial list.